dash/include/dash/algorithm/Sort.h

```diff
@@ -0,0 +1,1177 @@
+ * \c dash::sort must be global (\c GlobIter<ValueType>).
+ *
+ * The operation is collective among the team of the owning dash container.
+ *
+ * Example:
+ *
+ * \code
+ *       dash::Array<int> arr(100);
+ *       dash::generate(arr.begin(), arr.end());
+ *       dash::sort(array.begin(),
+ *                  array.end();
+ * \endcode
+ *
+ * \ingroup  DashAlgorithms
+ */
+template <class GlobRandomIt>
+void sort(GlobRandomIt begin, GlobRandomIt end);
+
+/**
+ * Sorts the elements in the range, defined by \c [begin, end) in ascending
+ * order. The order of equal elements is not guaranteed to be preserved.
+ *
+ * Elements are sorted by using a user-defined hash function. This hash
+ * function is required to be sortable by operator<.
+ *
+ * This variant may be appropriate if the underlying container does not hold
+ * arithmetic values (e.g. structs).
+ *
+ * In terms of data distribution, source and destination ranges passed to
+ * \c dash::sort must be global (\c GlobIter<ValueType>).
+ *
+ * The operation is collective among the team of the owning dash container.
+ *
+ * Example:
+ *
+ * \code
+ *       struct pair { int x; int y; };
+ *       dash::Array<pair> arr(100);
+ *       dash::generate(arr.begin(), arr.end(), random());
+ *       dash::sort(array.begin(),
+ *                  array.end(),
+ *                  [](pair const & p){return p.x % 13; }
+ *                 );
+ * \endcode
+ *
+ * \ingroup  DashAlgorithms
+ */
+template <class GlobRandomIt, class SortableHash>
+void sort(GlobRandomIt begin, GlobRandomIt end, SortableHash hash);
+
+#else
+
+#ifdef DASH_ENABLE_TRACE_LOGGING
+#define MAX_ELEMS_LOGGING 25
+#define DASH_SORT_LOG_TRACE_RANGE(desc, begin, end)                         \
+  do {                                                                      \
+    using value_t =                                                         \
+        typename std::iterator_traits<decltype(begin)>::value_type;         \
+    using difference_t =                                                    \
+        typename std::iterator_traits<decltype(begin)>::difference_type;    \
+    auto const nelems = std::distance(begin, end);                          \
+    auto const max_elems =                                                  \
+        std::min<difference_t>(nelems, MAX_ELEMS_LOGGING);                  \
+    std::ostringstream os;                                                  \
+    std::copy(                                                              \
+        begin, begin + max_elems, std::ostream_iterator<value_t>(os, " ")); \
+    if (nelems > MAX_ELEMS_LOGGING) os << "...";                            \
+    DASH_LOG_TRACE(desc, os.str());                                         \
+  } while (0)
+#else
+#define DASH_SORT_LOG_TRACE_RANGE(desc, begin, end) \
+  do {                                              \
+  } while (0)
+#endif
+
+#define IND_DIST(nunits) (nunits * 0)
```

Comment:
What is the semantic of these macros? Indizes? If so, better use `IDX_DIST`

Also: `((nunits) * 0)`, similar below